// angular-fusioncharts/src/fusioncharts.pipe.js
import { Injectable, Pipe } from '@angular/core';
/**
 * Transforms any input value
 */
import * as ɵngcc0 from '@angular/core';
var FusionChartsPipe = /** @class */ (function () {
    function FusionChartsPipe() {
    }
    FusionChartsPipe.prototype.transform = function (value, args) {
        return value;
    };
    /** @nocollapse */
    FusionChartsPipe.ctorParameters = function () { return []; };
FusionChartsPipe.ngInjectableDef = ɵngcc0.ɵɵdefineInjectable({ token: FusionChartsPipe, factory: function FusionChartsPipe_Factory(t) { return new (t || FusionChartsPipe)(); }, providedIn: null });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(FusionChartsPipe, [{
        type: Pipe,
        args: [{
                name: 'fusionchartsPipe'
            }]
    }, {
        type: Injectable
    }], function () { return []; }, { transform: [] });
FusionChartsPipe.ngPipeDef = ɵngcc0.ɵɵdefinePipe({ name: "fusionchartsPipe", type: FusionChartsPipe, factory: function FusionChartsPipe_Factory(t) { return new (t || FusionChartsPipe)(); }, pure: true });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(FusionChartsPipe, [{
        type: Pipe,
        args: [{
                name: 'fusionchartsPipe'
            }]
    }, {
        type: Injectable
    }], function () { return []; }, { transform: [] });
    return FusionChartsPipe;
}());
export { FusionChartsPipe };
