// agm-direction/esm2015/src/modules/agm-direction.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { AgmDirection } from '../directive/agm-direction.directive';
import * as ɵngcc0 from '@angular/core';
export { AgmDirection } from '../directive/agm-direction.directive';
export class AgmDirectionModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: AgmDirectionModule,
        };
    }
}
AgmDirectionModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: AgmDirectionModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(AgmDirectionModule, { declarations: function () { return [AgmDirection]; }, exports: function () { return [AgmDirection]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(AgmDirectionModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [
                    AgmDirection,
                ],
                exports: [
                    AgmDirection,
                ]
            }]
    }], null, null);
AgmDirectionModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function AgmDirectionModule_Factory(t) { return new (t || AgmDirectionModule)(); }, imports: [[]] });
