// @ckeditor/ckeditor5-angular/esm5/ckeditor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { CKEditorComponent } from './ckeditor.component';
import * as ɵngcc0 from '@angular/core';
var CKEditorModule = /** @class */ (function () {
    function CKEditorModule() {
    }
CKEditorModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: CKEditorModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(CKEditorModule, { declarations: function () { return [CKEditorComponent]; }, imports: function () { return [FormsModule, CommonModule]; }, exports: function () { return [CKEditorComponent]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(CKEditorModule, [{
        type: NgModule,
        args: [{
                imports: [FormsModule, CommonModule],
                declarations: [CKEditorComponent],
                exports: [CKEditorComponent]
            }]
    }], function () { return []; }, null);
CKEditorModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function CKEditorModule_Factory(t) { return new (t || CKEditorModule)(); }, imports: [[FormsModule, CommonModule]] });
    return CKEditorModule;
}());
export { CKEditorModule };
